src/home_robot/home_robot/agent/motion/pinocchio_ik_solver.py

```diff
@@ -0,0 +1,246 @@
+import os
+from typing import Tuple
+
+import numpy as np
+import pinocchio
+from scipy.spatial.transform import Rotation as R
+
+# Error tolerances
+POS_ERROR_TOL = 0.005
+ORI_ERROR_TOL = [0.1, 0.1, np.pi / 2]
+
+# CEM
+CEM_MAX_ITERATIONS = 5
+CEM_NUM_SAMPLES = 50
+CEM_NUM_TOP = 10
+
+# URDF
+HAB_STRETCH_PATH = "../hab_stretch"
+URDF_REL_PATH = "urdf/planner_calibrated_simplified.urdf"
+URDF_PATH = os.path.join(HAB_STRETCH_PATH, URDF_REL_PATH)
+
+EE_NAME = "link_straight_gripper"
+PIN_CONTROLLED_JOINTS = [
+    # "base_x_joint",
+    "joint_lift",
+    "joint_arm_l0",
+    "joint_arm_l1",
+    "joint_arm_l2",
+    "joint_arm_l3",
+    "joint_wrist_yaw",
+    "joint_wrist_pitch",
+    "joint_wrist_roll",
+]
+
+
+class PinocchioIKSolver:
+    EPS = 1e-4
+    DT = 1e-1
+    DAMP = 1e-12
+
+    def __init__(self, urdf_path, ee_link_name, controlled_joints):
+        self.model = pinocchio.buildModelFromUrdf(urdf_path)
+        self.data = self.model.createData()
+        self.q_neutral = pinocchio.neutral(self.model)
+
+        self.ee_frame_idx = [f.name for f in self.model.frames].index(ee_link_name)
+        self.controlled_joints = [
+            self.model.idx_qs[self.model.getJointId(j)] if j != "ignore" else -1
+            for j in controlled_joints
+        ]
+
+    def get_dof(self) -> int:
+        """returns dof for the manipulation chain"""
+        return len(self.controlled_joints)
+
+    def _qmap_control2model(self, q_input):
+        q_out = self.q_neutral.copy()
+        for i, joint_idx in enumerate(self.controlled_joints):
+            q_out[joint_idx] = q_input[i]
+
+        return q_out
+
+    def _qmap_model2control(self, q_input):
+        q_out = np.empty(len(self.controlled_joints))
+        for i, joint_idx in enumerate(self.controlled_joints):
+            if joint_idx >= 0:
+                q_out[i] = q_input[joint_idx]
+
+        return q_out
+
+    def compute_fk(self, q) -> Tuple[np.ndarray, np.ndarray]:
+        """given joint values return end-effector position and quaternion associated with it"""
+        q_model = self._qmap_control2model(q)
+        pinocchio.forwardKinematics(self.model, self.data, q_model)
+        pinocchio.updateFramePlacement(self.model, self.data, self.ee_frame_idx)
+        pos = self.data.oMf[self.ee_frame_idx].translation
+        quat = R.from_matrix(self.data.oMf[self.ee_frame_idx].rotation).as_quat()
```

Comment:
should return type be (pos, quat) like this, or are we using sophus?
cc @exhaustin 

for now best to be consistent with pybullet i guess

Reply:
What is "consistent with pybullet"?

Reply:
[pos, quat] is the pybullet standard: https://usermanual.wiki/Document/pybullet20quickstart20guide.479068914/html#pf17

I think we should unify pose representations across the entire repo but I don't have a very good candidate right now. Sophus is nice but it needs pybind to install which is slightly annoying. Maybe we can wait for the Theseus Lie library to release. Anyhow I think it's okay to just use pos, quat for now.